### PR TITLE
Add support for legacy amt v0

### PR DESCRIPTION
**Pull request by creativcoder**
This PR adds support for serializing and deserializing amt v0 data structures, which are still used in filecoin for reading headers etc.

Fixes: #924

### CODE REVIEW

**Comment by creativcoder** [comment]
* Using types would not allow us to have the pub type alias with compile time version declaration. 
  With types, we would need to expose a generic top level Amt. 
  With const version, I have type aliased the v3 to just Amt, and the change would only be affecting the v0 code.
* Using const we don't need to store an instance of version within the struct.
* Since version is only going to be an integral (in this particular case) i thought using const generic would be the simplest.
* With types we wouldn't be able to match it without an instance in serde methods.

Do we foresee any issues long term with this approach?

**Comment by Stebalien** [comment]
> Using types would not allow us to have the pub type alias with compile time version declaration.
With types, we would need to expose a generic top level Amt.
With const version, I have type aliased the v3 to just Amt, and the change would only be affecting the v0 code.


Ah, I mean replace the consts with unit structs (possibly implementing some trait. I.e.:


```rust
pub struct V0;
pub struct V3;

pub trait AmtVersion {} // ideally private..., but I'm pretty sure rust won't let us do that?

impl AmtVersion for V0 {}
impl AmtVersion for V3 {}

pub struct AmtImpl<V, BS, VER: AmtVersion> { ... }
```

It's _basically_ identical to the const version. The main differences are:

- We can use associated types in the type signature. I.e. `AmtImpl<Thing, MyBS, OtherType::Version>`. IIRC, this isn't stable for consts.
- We can restrict the allowed versions. Right now, someone can technically specify a version that makes no sense and we don't really check it anywhere.
- Given different future "features", we can add traits for these "features" and implement these traits on the versions that support them. Then we can say things like `AmtImpl<Thing, MyBS, V: HasSomeFeature>`.
- We can implement amt-version specific logic on the _version_ type itself. E.g., add a serialize/deserialize method.

On the other hand, the main benefit of the const is that it's easier to write "runtime" checks that check the const, that actually disappear at compile time (basically, it gives you specialization without specialization).

Basically, I just want to make sure you've seen and considered both options before proceeding. There are tradeoffs both ways and I haven't fully considered either option.

**Comment by creativcoder** [comment]
Yeah, you're right about being able to use any integer value for the version. Using traits with unit structs, is definitely extensible. I'm gonna try that, and update soon.

**Comment by creativcoder** [comment]
@mriise :eyes:  This is amazing! thanks for sharing.

File: ipld/amt/src/amt.rs [hunk 1]
```diff
@@ -37,32 +44,31 @@ use crate::{
...
-pub struct Amt<V, BS> {
-    root: Root<V>,
-    block_store: BS,
-}
+pub type Amt<V, BS> = AmtImpl<V, BS, V3>;
+/// Legacy amt V0
+pub type Amtv0<V, BS> = AmtImpl<V, BS, V0>;
 
-impl<V: PartialEq, BS: Blockstore> PartialEq for Amt<V, BS> {
+impl<V: PartialEq, BS: Blockstore, Ver: PartialEq> PartialEq for AmtImpl<V, BS, Ver> {
```
Comment (Stebalien) [inline ipld/amt/src/amt.rs:51]: We shouldn't _need_ PartialEq for `Ver`, right?

File: ipld/amt/src/amt.rs [hunk 2]
```diff
@@ -37,32 +44,31 @@ use crate::{
...
     pub fn new(block_store: BS) -> Self {
         Self::new_with_bit_width(block_store, DEFAULT_BIT_WIDTH)
     }
 
-    /// Construct new Amt with given bit width.
+    /// Construct new Amt with given bit width (v3)
     pub fn new_with_bit_width(block_store: BS, bit_width: u32) -> Self {
         Self {
-            root: Root::new(bit_width),
+            root: RootImpl::<V, Ver>::new_with_bit_width(bit_width),
```
Comment (Stebalien) [inline ipld/amt/src/amt.rs:71]: Is this type not inferred?

File: ipld/amt/src/root.rs [hunk 3]
```diff
@@ -1,62 +1,126 @@
...
+    const NUMBER: usize;
+}
+
+impl Version for V0 {
+    const NUMBER: usize = 0;
+}
+
+impl Version for V3 {
+    const NUMBER: usize = 3;
+}
```
Comment (Stebalien) [inline ipld/amt/src/root.rs:27]: I'd put these in a `version` submodule.

File: ipld/amt/src/amt.rs [hunk 4]
```diff
@@ -37,32 +44,31 @@ use crate::{
...
+    Ver: AmtVersion,
 {
-    /// Constructor for Root AMT node
+    /// Constructor for Root AMT node (v0)
     pub fn new(block_store: BS) -> Self {
         Self::new_with_bit_width(block_store, DEFAULT_BIT_WIDTH)
     }
 
-    /// Construct new Amt with given bit width.
+    /// Construct new Amt with given bit width (v3)
```
Comment (Stebalien) [inline ipld/amt/src/amt.rs:68]: These constructors aren't related to the version.

**Review by Stebalien — COMMENTED** [review]
Mostly nits at this point.

File: ipld/amt/src/root.rs [hunk 5]
```diff
@@ -1,62 +1,126 @@
...
+use crate::{init_sized_vec, Node, DEFAULT_BIT_WIDTH};
+
+#[derive(PartialEq, Debug)]
+pub struct V0;
+#[derive(PartialEq, Debug)]
+pub struct V3;
+
+pub trait Version {
+    const NUMBER: usize;
+}
```
Comment (mriise) [inline ipld/amt/src/root.rs:19]: I would be careful to fully expose this trait, as now users could implement their own versions and get trapped on unreachable if they ever use it. I would either static assert the version, or make it pub(crate) or something.

File: ipld/amt/src/amt.rs (hunk 1 above)
Comment (creativcoder) [reply ipld/amt/src/amt.rs:51]: It shouldn't, I too found it weird but here's the error:

```
error[E0369]: binary operation `==` cannot be applied to type `RootImpl<V, Ver>`
  --> ipld/amt/src/amt.rs:53:19
   |
53 |         self.root == other.root
   |         --------- ^^ ---------- RootImpl<V, Ver>
   |         |
   |         RootImpl<V, Ver>
   |
help: consider restricting type parameter `Ver`
   |
51 | impl<V: PartialEq, BS: Blockstore, Ver: std::cmp::PartialEq> PartialEq for AmtImpl<V, BS, Ver> {
```

File: ipld/amt/src/amt.rs (hunk 2 above)
Comment (creativcoder) [reply ipld/amt/src/amt.rs:71]: Ah it does, removed turbofishes.

File: ipld/amt/src/root.rs (hunk 3 above)
Comment (creativcoder) [reply ipld/amt/src/root.rs:27]: added

File: ipld/amt/src/amt.rs (hunk 4 above)
Comment (creativcoder) [reply ipld/amt/src/amt.rs:68]: removed

File: ipld/amt/src/root.rs (hunk 5 above)
Comment (creativcoder) [reply ipld/amt/src/root.rs]: I have moved it behind a version submodule and pub(crate)'d it.

**Comment by creativcoder** [comment]
btw do we require a minor version bump for this?

**Comment by Stebalien** [comment]
> btw do we require a minor version bump for this?

It looks completely backwards compatible, so no.

File: ipld/amt/src/amt.rs (hunk 1 above)
Comment (Stebalien) [reply ipld/amt/src/amt.rs:52]: Hm. Ok, so, I think that's just a derive issue (you could manually implement it). But honestly, it doesn't really matter.

File: ipld/amt/src/amt.rs [hunk 6]
```diff
@@ -12,10 +12,18 @@ use itertools::sorted;
...
 use crate::node::{CollapsedNode, Link};
+use crate::root::version::{Version as AmtVersion, V0, V3};
+use crate::root::RootImpl;
 use crate::{
-    init_sized_vec, nodes_for_height, Error, Node, Root, DEFAULT_BIT_WIDTH, MAX_HEIGHT, MAX_INDEX,
+    init_sized_vec, nodes_for_height, Error, Node, DEFAULT_BIT_WIDTH, MAX_HEIGHT, MAX_INDEX,
 };
 
+#[derive(Debug)]
+pub struct AmtImpl<V, BS, Ver> {
```
Comment (Stebalien) [inline ipld/amt/src/amt.rs:22]: Can we add:

```rust
#[doc(hidden)]
```

File: ipld/amt/src/amt.rs [hunk 7]
```diff
@@ -37,32 +45,31 @@ use crate::{
...
 
-impl<V, BS> Amt<V, BS>
+impl<V, BS, Ver> AmtImpl<V, BS, Ver>
 where
     V: DeserializeOwned + Serialize,
     BS: Blockstore,
+    Ver: AmtVersion,
 {
-    /// Constructor for Root AMT node
+    /// Constructor for Root AMT node (v0)
```
Comment (Stebalien) [inline ipld/amt/src/amt.rs:64]: Not v0 specific.

File: ipld/amt/src/root.rs [hunk 8]
```diff
@@ -1,62 +1,128 @@
...
             node: Node::Leaf {
                 vals: init_sized_vec(bit_width),
             },
+            ver: PhantomData,
+        }
+    }
+}
+
+impl<V> RootImpl<V, self::version::V0> {
+    pub(crate) fn new() -> Rootv0<V> {
```
Comment (Stebalien) [inline ipld/amt/src/root.rs:60]: Let's make the visibility of `new` and `new_with_bit_width` match (just pick the same one, either one).

**Review by Stebalien — COMMENTED** [review]
Some super-nits but this is otherwise ready to merge.

**Comment by Stebalien** [comment]
Hm. Except that there are some clippy lints and warnings.

File: ipld/amt/src/root.rs (hunk 8 above)
Comment (creativcoder) [reply ipld/amt/src/root.rs:60]: Sure, using `pub(crate)`

File: ipld/amt/src/amt.rs (hunk 6 above)
Comment (creativcoder) [reply ipld/amt/src/amt.rs:22]: Added

File: ipld/amt/src/amt.rs (hunk 7 above)
Comment (creativcoder) [reply ipld/amt/src/amt.rs:64]: Fixed

**Comment by creativcoder** [comment]
I'm not sure why there's still a warning on type alias of RootImpl as I did use them both in [root.rs](https://github.com/creativcoder/ref-fvm/blob/913751785914eac44db08a17174b00421872d696/ipld/amt/src/root.rs#L158)

**Comment by Stebalien** [comment]
Ah, well, tests don't count. I'd just define these aliases inside the tests.

**Comment by creativcoder** [comment]
Ah I overlooked the point that Root isn't meant to be part of public API. Moved the type aliases.

**Comment by Stebalien** [comment]
Thanks!
